onnxruntime/core/optimizer/selectors_actions/helpers.h

```diff
@@ -5,20 +5,49 @@
 
 namespace onnxruntime {
 
+// if the last input in num_inputs is for the variadic input, the variadic input could have zero or more values
+// so we need to special case the zero and count that as one. same for outputs
+inline int NumIOEntries(bool variadic_io, int num_io, int num_variadic_io) {
+  return variadic_io
+             ? num_io + std::max(1, num_variadic_io) - 1
+             : num_io;
+}
+
 //
 // Selection helpers
 //
 
-// Struct to serialize the node indexes in an ORT format model.
+// Struct to serialize the node indices in an ORT format model.
 // Use EmptyNodeIndex for nullptr entries in the vectors for missing optional inputs
-struct NodesToOptimizeIndexes {
-  std::vector<NodeIndex> nodes;
-  int num_inputs;
-  int num_outputs;
-  bool variadic_input;
-  bool variadic_output;
-  int num_variadic_inputs;
-  int num_variadic_outputs;
+struct NodesToOptimizeIndices {
+  const std::vector<NodeIndex> nodes_;
+  const int num_inputs_;
+  const int num_outputs_;
+  const bool variadic_input_;
+  const bool variadic_output_;
+  const int num_variadic_inputs_;
+  const int num_variadic_outputs_;
+
+  NodesToOptimizeIndices(std::vector<NodeIndex>&& nodes,
+                         int num_inputs,
+                         int num_outputs,
+                         bool variadic_input,
+                         bool variadic_output,
+                         int num_variadic_inputs,
+                         int num_variadic_outputs)
+      : nodes_(std::move(nodes)),
+        num_inputs_(num_inputs),
+        num_outputs_(num_outputs),
+        variadic_input_(variadic_input),
+        variadic_output_(variadic_output),
+        num_variadic_inputs_(num_variadic_inputs),
+        num_variadic_outputs_(num_variadic_outputs) {}
```

Comment:
could we add some logic to ensure that node index values are less than EmptyNodeIndex? that got changed from `std::numeric_limits<NodeIndex>::max()` to `std::numeric_limits<uint32_t>::max()` in #9508. creation of NodesToOptimizeIndices might be a good place for it.

also, sorry about causing a bunch of merge conflicts

Reply:
I get rid of the ctor, added the validation to `GetNodesToOptimizeIndices`